File: apps/mantui/src/cli/arguments.ts
/**
 * @file Parses the interactive `mantui` command line without loading the TUI.
 *
 * Non-interactive outline, text, JSON, Markdown, and cache operations belong
 * to the separate Rust `mant`, leaving this command focused on TUI use.
 */

// ── Public command model ───────────────────────────────────────────────────

export type CliCommand =
  | { kind: "help" }
  | {
    kind: "query";
    input:
      | { kind: "manual"; topic: string; section?: string }
      | { kind: "markdown-file"; path: string };
    forceLibmandoc?: boolean;
    forceGroff?: boolean;
  };

/** An invalid invocation; callers report these with exit code 2. */
export class CliUsageError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CliUsageError";
  }
}

// ── Help text ──────────────────────────────────────────────────────────────

export const CLI_HELP = `ManT — explore local manuals and Markdown in a structured terminal UI

Usage:
  mantui <topic|markdown> [--section <section>] [--force-libmandoc] [--force-groff]
  mantui -h, --help

Options:
  -h, --help              Show this help and exit
  -s, --section <value>   Select a manual section, such as 1 or 3p
  --force-libmandoc       Require direct libmandoc output and print diagnostics
  --force-groff           Use man -Thtml + groff HTML parser instead of libmandoc
                           (not comprehensively tested)
  --                      Treat all remaining arguments as the topic

What ManT provides:
  Structured documents with a hierarchy-aware sidebar, in-page references,
  document search, and optional or embedded tldr quick references.

For agents and scripts:
  mant <topic> --outline               Discover sections and options
  mant <topic> --explain=--option      Explain one semantic entry
  mant <topic> --search=<pattern>      Find text with stable locations
  mant -h                              Show Markdown, text, JSON, and schema output

Examples:
  mantui git
  mantui README.md
  mantui printf --section 3
  mantui --force-libmandoc tar`;

// ── Parser ─────────────────────────────────────────────────────────────────

/** Converts raw argv tokens into one validated interactive command. */
export function parseCliArguments(args: readonly string[]): CliCommand {
  let showHelp = false;
  let parseOptions = true;
  let section: string | undefined;
  let forceLibmandoc = false;
  let forceGroff = false;
  const topicParts: string[] = [];

  for (let index = 0; index < args.length; index++) {
    const arg = args[index]!;
    if (parseOptions && arg === "--") {
      parseOptions = false;
    } else if (parseOptions && (arg === "--help" || arg === "-h")) {
      showHelp = true;
    } else if (parseOptions && (arg === "--section" || arg === "-s")) {
      const value = args[++index];
      if (value === undefined) throw new CliUsageError("--section requires a value");
      if (section !== undefined) {
        throw new CliUsageError("--section may only be supplied once");
      }
      section = value;
    } else if (parseOptions && arg === "--force-libmandoc") {
      forceLibmandoc = true;
    } else if (parseOptions && arg === "--force-groff") {
      forceGroff = true;
    } else if (parseOptions && arg.startsWith("-")) {
      throw new CliUsageError(
        `unknown option '${arg}'; non-interactive output is provided by mant`,
      );
    } else {
      topicParts.push(arg);
    }
  }

  // Help stays side-effect free even when another token was supplied.
  if (showHelp) return { kind: "help" };

  const value = topicParts.join(" ").trim();
  if (!value) throw new CliUsageError("a manual topic or Markdown path is required");
  if (forceLibmandoc && forceGroff) {
    throw new CliUsageError("--force-libmandoc and --force-groff cannot be used together");
  }
  if (section !== undefined && !section.trim()) {
    throw new CliUsageError("manual section must not be empty");
  }
  const markdown = isMarkdownPath(value);
  if (markdown && section !== undefined) {
    throw new CliUsageError("--section applies only to manual topics");
  }
  if (markdown && (forceLibmandoc || forceGroff)) {
    throw new CliUsageError("manual renderer policies do not apply to Markdown input");
  }
  return {
    kind: "query",
    input: markdown
      ? { kind: "markdown-file", path: value }
      : {
          kind: "manual",
          topic: value,
          ...(section === undefined ? {} : { section: section.trim() }),
        },
    ...(forceLibmandoc ? { forceLibmandoc: true } : {}),
    ...(forceGroff ? { forceGroff: true } : {}),
  };
}

function isMarkdownPath(value: string): boolean {
  const lower = value.toLocaleLowerCase();
  return lower.endsWith(".md")
    || lower.endsWith(".markdown")
    || value.startsWith(".")
    || value.includes("/")
    || value.includes("\\");
}

File: engine/crates/mant-ui/src/arguments.rs
//! Defines the temporary Rust TUI command line while it remains a separate binary.

use std::path::Path;

use clap::{CommandFactory, Parser, error::ErrorKind};
use mant_ast::QueryInput;
use mant_core::QueryPolicy;

/// Inputs accepted by the interactive document reader.
#[derive(Debug, Parser)]
#[command(
    name = "mantui-rs",
    about = "Explore local manuals and Markdown in a structured terminal UI",
    long_about = "Explore complete local Unix manuals and Markdown documents with a hierarchy-aware sidebar, full-page search, page-local references, and optional tldr quick references.",
    after_help = "Examples:\n  mantui-rs git\n  mantui-rs README.md\n  mantui-rs printf --section 3\n  mantui-rs --force-libmandoc tar"
)]
pub(crate) struct Arguments {
    /// Manual topic or local Markdown path.
    #[arg(required = true, num_args = 1.., value_name = "TOPIC|MARKDOWN")]
    input: Vec<String>,

    /// Select a manual section such as 1 or 3p.
    #[arg(short, long, value_name = "SECTION")]
    section: Option<String>,

    /// Require direct libmandoc output and print its diagnostics.
    #[arg(long, conflicts_with = "force_groff")]
    force_libmandoc: bool,

    /// Use man -Thtml and the groff HTML parser.
    #[arg(long)]
    force_groff: bool,
}

/// A validated query and its host-only parser policy.
#[derive(Debug)]
pub(crate) struct Invocation {
    pub(crate) input: QueryInput,
    pub(crate) policy: QueryPolicy,
}

impl Arguments {
    pub(crate) fn invocation(self) -> Result<Invocation, clap::Error> {
        let value = self.input.join(" ");
        let value = value.trim();
        if value.is_empty() {
            return Err(Self::command().error(
                ErrorKind::InvalidValue,
                "a manual topic or Markdown path is required",
            ));
        }

        let markdown = is_markdown_path(value);
        if markdown && self.section.is_some() {
            return Err(Self::command().error(
                ErrorKind::ArgumentConflict,
                "--section applies only to manual topics",
            ));
        }
        if markdown && (self.force_libmandoc || self.force_groff) {
            return Err(Self::command().error(
                ErrorKind::ArgumentConflict,
                "manual renderer policies do not apply to Markdown input",
            ));
        }

        let section = self.section.map(|section| section.trim().to_owned());
        if section.as_deref() == Some("") {
            return Err(
                Self::command().error(ErrorKind::InvalidValue, "manual section must not be empty")
            );
        }

        Ok(Invocation {
            input: if markdown {
                QueryInput::MarkdownFile {
                    path: value.to_owned(),
                }
            } else {
                QueryInput::Manual {
                    topic: value.to_owned(),
                    section,
                }
            },
            policy: QueryPolicy {
                force_libmandoc: self.force_libmandoc,
                force_groff: self.force_groff,
            },
        })
    }
}

fn is_markdown_path(input: &str) -> bool {
    let path = Path::new(input);
    path.is_file()
        || path.extension().is_some_and(|extension| {
            extension.eq_ignore_ascii_case("md") || extension.eq_ignore_ascii_case("markdown")
        })
        || input.starts_with('.')
        || input.contains('/')
        || input.contains('\\')
}

#[cfg(test)]
mod tests {
    use clap::Parser;
    use mant_ast::QueryInput;
    use mant_core::QueryPolicy;

    use super::Arguments;

    fn parse(values: &[&str]) -> super::Invocation {
        Arguments::try_parse_from(values)
            .expect("parse arguments")
            .invocation()
            .expect("validate arguments")
    }

    #[test]
    fn parses_manual_sections_and_parser_policies() {
        let invocation = parse(&["mantui-rs", "printf", "--section", "3"]);
        assert_eq!(
            invocation.input,
            QueryInput::Manual {
                topic: "printf".to_owned(),
                section: Some("3".to_owned()),
            }
        );
        assert_eq!(invocation.policy, QueryPolicy::default());

        let invocation = parse(&["mantui-rs", "--force-libmandoc", "tar"]);
        assert!(invocation.policy.force_libmandoc);
        assert!(!invocation.policy.force_groff);
    }

    #[test]
    fn identifies_markdown_paths_without_touching_the_file_system() {
        for path in ["README.md", "guide.markdown", "docs/guide", r"docs\guide"] {
            assert_eq!(
                parse(&["mantui-rs", path]).input,
                QueryInput::MarkdownFile {
                    path: path.to_owned(),
                }
            );
        }
    }

    #[test]
    fn accepts_a_dash_prefixed_topic_after_the_option_terminator() {
        assert_eq!(
            parse(&["mantui-rs", "--", "-topic"]).input,
            QueryInput::Manual {
                topic: "-topic".to_owned(),
                section: None,
            }
        );
    }

    #[test]
    fn rejects_conflicting_or_inapplicable_options() {
        assert!(
            Arguments::try_parse_from(["mantui-rs", "tar", "--force-libmandoc", "--force-groff"])
                .is_err()
        );

        let markdown = Arguments::try_parse_from(["mantui-rs", "README.md", "--section", "1"])
            .expect("syntactically valid")
            .invocation()
            .expect_err("section does not apply to Markdown");
        assert!(markdown.to_string().contains("only to manual topics"));
    }
}
